grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -476,24 +484,74 @@ public PickResult pickSubchannel(PickSubchannelArgs args) {
   }
 
   @VisibleForTesting
+  static final class RoundRobinEntry {
+    final PickResult result;
+    @Nullable
+    final String token;
+
+    /**
+     * A non-drop result.
+     */
+    RoundRobinEntry(Subchannel subchannel, String token) {
+      this.result = PickResult.withSubchannel(subchannel);
+      this.token = token;
+    }
+
+    /**
+     * A drop result.
+     */
+    RoundRobinEntry(Status dropStatus) {
+      this.result = PickResult.withError(dropStatus);
+      this.token = null;
+    }
+
+    void updateHeaders(Metadata headers) {
+      if (token != null) {
+        headers.discardAll(TOKEN_KEY);
+        headers.put(TOKEN_KEY, token);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return "[result=" + result + ", token=" + token + "]";
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(result, token);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (!(other instanceof RoundRobinEntry)) {
+        return false;
+      }
+      RoundRobinEntry that = (RoundRobinEntry) other;
+      return Objects.equal(result, that.result) && Objects.equal(token, that.token);
+    }
+  }
+
+  @VisibleForTesting
   static final class RoundRobinPicker extends SubchannelPicker {
-    final List<PickResult> list;
+    final List<RoundRobinEntry> list;
     int index;
 
-    RoundRobinPicker(List<PickResult> resultList) {
+    RoundRobinPicker(List<RoundRobinEntry> resultList) {
       checkArgument(!resultList.isEmpty(), "resultList is empty");
       list = resultList;
     }
 
     @Override
     public PickResult pickSubchannel(PickSubchannelArgs args) {
       synchronized (list) {
-        PickResult result = list.get(index);
+        RoundRobinEntry result = list.get(index);
         index++;
         if (index == list.size()) {
           index = 0;
         }
-        return result;
+        result.updateHeaders(args.getHeaders());
```

Comment:
This is intended. The `args` doesn't only pass information to the picker, but also pass mutable objects like headers for the picker to mutate. You don't see such pattern very often because most objects in gRPC API are immutable, while `Metadata` is an exception. Whenever it comes to changing `Metadata`, it is passed in as argument instead of being returned. Another example is `CallCredentials.MetadataApplier.apply(Metadata headers)`.